teos/src/api/lightning.rs

```diff
@@ -0,0 +1,965 @@
+pub struct Logger;
+
+impl LightningLogger for Logger {
+    fn log(&self, record: &Record) {
+        #[cfg(test)]
+        // Pass "-- --nocapture" flag to "cargo test" for this println to appear.
+        println!(
+            // "\x1B" stuff are terminal colors. Might not work in some terminals though.
+            "\x1B[42m{}\x1B[0m [\x1B[33m{}:{}\x1B[0m]: {}",
+            record.level, record.module_path, record.line, record.args
+        );
+        #[cfg(not(test))]
+        match record.level {
+            Level::Error => log::error!(target: record.module_path, "{}", record.args),
+            Level::Warn => log::warn!(target: record.module_path, "{}", record.args),
+            Level::Info => log::info!(target: record.module_path, "{}", record.args),
+            Level::Debug => log::debug!(target: record.module_path, "{}", record.args),
+            Level::Trace => log::trace!(target: record.module_path, "{}", record.args),
+            _ => {}
+        }
+    }
+}
+
+pub async fn serve(
+    lightning_bind: SocketAddr,
+    grpc_bind: String,
+    shutdown_signal: Listener,
+    tower_sk: SecretKey,
+) {
+    let grpc_conn = loop {
+        match PublicTowerServicesClient::connect(grpc_bind.clone()).await {
+            Ok(conn) => break conn,
+            Err(_) => {
+                log::error!("Cannot connect to the gRPC server. Retrying shortly");
+                tokio::time::sleep(tokio::time::Duration::from_secs(1)).await;
+            }
+        }
+    };
+    let tower_message_handler = Arc::new(TowerMessageHandler::new(
+        grpc_conn,
+        tokio::runtime::Handle::current(),
+    ));
+    let message_handler = MessageHandler {
+        chan_handler: Arc::new(ErroringMessageHandler::new()),
+        route_handler: Arc::new(IgnoringMessageHandler {}),
+    };
+    let ephemeral_bytes: [u8; 32] = get_random_bytes(32).try_into().unwrap();
+    let peer_manager = Arc::new(TowerPeerManager::new(
+        message_handler,
+        tower_sk,
+        &ephemeral_bytes,
+        Arc::new(Logger),
+        tower_message_handler,
+    ));
+    // To suppress an issue similar to https://github.com/rust-lang/rust-clippy/issues/2928
+    #[allow(clippy::expect_fun_call)]
+    let listener = tokio::net::TcpListener::bind(lightning_bind)
+        .await
+        .expect(&format!(
+            "Couldn't bind the lightning server to {}",
+            lightning_bind
+        ));
+    // A tokio runtime handle to run lightning net tokio on. This is to fix a deadlock issue similar
+    // to https://github.com/lightningdevkit/rust-lightning/issues/1367 which appears with too many
+    // concurrent requests to the server.
+    let ldk_handle = Box::leak(Box::new(tokio::runtime::Runtime::new().unwrap()))
+        .handle()
+        .clone();
+    loop {
+        let tcp_stream = listener.accept().await.unwrap().0;
+        if shutdown_signal.is_triggered() {
+            return;
+        }
```

Comment:
This actually waits until there is a new inbound connection then checks for if a shutdown signal was issued.
We need to listen for inbound and shutdown signal at the same time.